File: packages/protocol/src/room.ts
import * as decoding from 'lib0/decoding'
import * as encoding from 'lib0/encoding'
import * as awarenessProtocol from 'y-protocols/awareness'
import * as syncProtocol from 'y-protocols/sync'
import type * as Y from 'yjs'
import { decrypt, encrypt } from './cipher.ts'
import { decodeMessage, encodeMessage, MessageType } from './message.ts'

export type RoomStatus = 'connecting' | 'connected' | 'disconnected'

/** The subset of the WebSocket API used by RoomClient (browser and Node.js 22+ globals both fit). */
export interface SocketLike {
  binaryType: string
  readonly readyState: number
  onopen: ((ev: unknown) => void) | null
  onmessage: ((ev: { data: unknown }) => void) | null
  onclose: ((ev: unknown) => void) | null
  onerror: ((ev: unknown) => void) | null
  send(data: Uint8Array): void
  close(code?: number, reason?: string): void
}

export interface RoomClientOptions {
  /** WebSocket URL of the room. Must never contain the key. */
  url: string
  key: CryptoKey
  doc: Y.Doc
  awareness: awarenessProtocol.Awareness
  createSocket?: (url: string) => SocketLike
  minBackoffMs?: number
  maxBackoffMs?: number
  onStatus?: (status: RoomStatus) => void
  onError?: (error: unknown) => void
}

const OPEN = 1

/**
 * Keeps a Y.Doc and Awareness in sync with every other peer in a room.
 *
 * The server is a dumb relay, so the peers run a symmetric protocol: on connect a
 * peer sends SyncStep1 (to pull what it is missing), its full state as an update
 * (to push what others are missing) and its awareness state. Every peer answers
 * SyncStep1 with SyncStep2. All frames are end-to-end encrypted.
 */
export class RoomClient {
  readonly doc: Y.Doc
  readonly awareness: awarenessProtocol.Awareness
  status: RoomStatus = 'disconnected'

  private readonly opts: RoomClientOptions
  private socket: SocketLike | null = null
  private attempts = 0
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null
  private destroyed = false
  private sendChain: Promise<void> = Promise.resolve()
  private receiveChain: Promise<void> = Promise.resolve()

  constructor(opts: RoomClientOptions) {
    this.opts = opts
    this.doc = opts.doc
    this.awareness = opts.awareness
    this.doc.on('update', this.handleDocUpdate)
    this.awareness.on('update', this.handleAwarenessUpdate)
    this.awareness.on('change', this.handleAwarenessChange)
  }

  connect(): void {
    if (this.destroyed || this.socket) return
    this.setStatus('connecting')
    const createSocket =
      this.opts.createSocket ?? ((url: string) => new WebSocket(url) as unknown as SocketLike)
    const socket = createSocket(this.opts.url)
    socket.binaryType = 'arraybuffer'
    this.socket = socket

    socket.onopen = () => {
      if (this.socket !== socket) return
      this.attempts = 0
      this.setStatus('connected')
      this.sendSync((encoder) => syncProtocol.writeSyncStep1(encoder, this.doc))
      this.sendSync((encoder) => syncProtocol.writeSyncStep2(encoder, this.doc))
      this.sendAwareness([this.doc.clientID])
    }
    socket.onmessage = (ev) => {
      if (this.socket !== socket) return
      const data = toBytes(ev.data)
      if (!data) return
      this.receiveChain = this.receiveChain
        .then(() => this.receive(data))
        .catch((error) => this.opts.onError?.(error))
    }
    socket.onclose = () => {
      if (this.socket !== socket) return
      this.socket = null
      this.dropRemoteAwareness()
      this.setStatus('disconnected')
      this.scheduleReconnect()
    }
    socket.onerror = (error) => {
      this.opts.onError?.(error)
    }
  }

  /** Announces departure, closes the socket and stops reconnecting. */
  async destroy(): Promise<void> {
    if (this.destroyed) return
    awarenessProtocol.removeAwarenessStates(this.awareness, [this.doc.clientID], 'local')
    await this.sendChain
    this.destroyed = true
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer)
    this.doc.off('update', this.handleDocUpdate)
    this.awareness.off('update', this.handleAwarenessUpdate)
    this.awareness.off('change', this.handleAwarenessChange)
    const socket = this.socket
    this.socket = null
    socket?.close()
    this.setStatus('disconnected')
  }

  private async receive(data: Uint8Array): Promise<void> {
    const message = decodeMessage(await decrypt(this.opts.key, data))
    const decoder = decoding.createDecoder(message.payload)
    if (message.type === MessageType.Sync) {
      const encoder = encoding.createEncoder()
      syncProtocol.readSyncMessage(decoder, encoder, this.doc, this)
      if (encoding.length(encoder) > 0) {
        this.send(MessageType.Sync, encoding.toUint8Array(encoder))
      }
    } else {
      awarenessProtocol.applyAwarenessUpdate(
        this.awareness,
        decoding.readVarUint8Array(decoder),
        this,
      )
    }
  }

  private handleDocUpdate = (update: Uint8Array, origin: unknown) => {
    if (origin === this) return
    this.sendSync((encoder) => syncProtocol.writeUpdate(encoder, update))
  }

  private handleAwarenessUpdate = (
    { added, updated, removed }: AwarenessChanges,
    origin: unknown,
  ) => {
    if (origin === this) return
    this.sendAwareness([...added, ...updated, ...removed])
  }

  // Someone new showed up: tell them about us right away instead of waiting for the
  // periodic awareness renewal.
  private handleAwarenessChange = ({ added }: AwarenessChanges, origin: unknown) => {
    if (origin === this && added.length > 0) this.sendAwareness([this.doc.clientID])
  }

  private dropRemoteAwareness(): void {
    const remote = [...this.awareness.getStates().keys()].filter((id) => id !== this.doc.clientID)
    awarenessProtocol.removeAwarenessStates(this.awareness, remote, this)
  }

  private sendSync(write: (encoder: encoding.Encoder) => void): void {
    const encoder = encoding.createEncoder()
    write(encoder)
    this.send(MessageType.Sync, encoding.toUint8Array(encoder))
  }

  private sendAwareness(clients: number[]): void {
    const encoder = encoding.createEncoder()
    encoding.writeVarUint8Array(
      encoder,
      awarenessProtocol.encodeAwarenessUpdate(this.awareness, clients),
    )
    this.send(MessageType.Awareness, encoding.toUint8Array(encoder))
  }

  // Encryption is async; chain sends so frames leave in the order they were produced.
  private send(type: MessageType, payload: Uint8Array): void {
    const socket = this.socket
    if (!socket || socket.readyState !== OPEN) return
    const plaintext = encodeMessage(type, payload)
    this.sendChain = this.sendChain
      .then(async () => {
        const data = await encrypt(this.opts.key, plaintext)
        if (socket.readyState === OPEN) socket.send(data)
      })
      .catch((error) => this.opts.onError?.(error))
  }

  private scheduleReconnect(): void {
    if (this.destroyed) return
    const min = this.opts.minBackoffMs ?? 500
    const max = this.opts.maxBackoffMs ?? 30_000
    const delay = Math.min(max, min * 2 ** this.attempts) * (0.5 + Math.random() / 2)
    this.attempts++
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null
      this.connect()
    }, delay)
  }

  private setStatus(status: RoomStatus): void {
    if (this.status === status) return
    this.status = status
    this.opts.onStatus?.(status)
  }
}

interface AwarenessChanges {
  added: number[]
  updated: number[]
  removed: number[]
}

function toBytes(data: unknown): Uint8Array | null {
  if (data instanceof ArrayBuffer) return new Uint8Array(data)
  if (ArrayBuffer.isView(data)) return new Uint8Array(data.buffer, data.byteOffset, data.byteLength)
  return null
}

File: internal/protocol/room.go
package protocol

import (
	"context"
	"math/rand/v2"
	"net/http"
	"sync"
	"time"

	"github.com/reearth/ygo/awareness"
	"github.com/reearth/ygo/crdt"
	"github.com/reearth/ygo/encoding"
	ysync "github.com/reearth/ygo/sync"
)

// Status is the connection state of a Client. StatusClosed is final: the host
// ended the session.
type Status string

const (
	StatusConnecting   Status = "connecting"
	StatusConnected    Status = "connected"
	StatusDisconnected Status = "disconnected"
	StatusClosed       Status = "closed"
)

type ClientOptions struct {
	// URL is the WebSocket URL of the room. It must never contain the key.
	URL       string
	Key       []byte
	Doc       *crdt.Doc
	Awareness *awareness.Awareness
	// Header holds extra handshake headers.
	Header     http.Header
	Dial       Dialer
	MinBackoff time.Duration
	MaxBackoff time.Duration
	OnStatus   func(Status)
	OnError    func(error)
}

// Client keeps a Doc and Awareness in sync with every other peer in a room.
//
// The server is a dumb relay, so the peers run a symmetric protocol: on connect a
// peer sends SyncStep1 (to pull what it is missing), its full state as SyncStep2
// (to push what others are missing) and its awareness state. Every peer answers
// SyncStep1 with SyncStep2. All frames are end-to-end encrypted.
//
// Changes applied from the room carry the Client as their origin.
type Client struct {
	opts   ClientOptions
	cipher *Cipher
	doc    *crdt.Doc
	aw     *awareness.Awareness

	// applying serializes applying remote updates with Do.
	applying sync.Mutex

	mu        sync.Mutex
	status    Status
	out       *outbox
	started   bool
	destroyed bool
	unsubs    []func()

	ctx    context.Context
	cancel context.CancelFunc
	done   chan struct{}
}

func NewClient(opts ClientOptions) (*Client, error) {
	cipher, err := NewCipher(opts.Key)
	if err != nil {
		return nil, err
	}
	if opts.Dial == nil {
		opts.Dial = DialWebSocket
	}
	if opts.MinBackoff == 0 {
		opts.MinBackoff = 500 * time.Millisecond
	}
	if opts.MaxBackoff == 0 {
		opts.MaxBackoff = 30 * time.Second
	}
	ctx, cancel := context.WithCancel(context.Background())
	c := &Client{
		opts:   opts,
		cipher: cipher,
		doc:    opts.Doc,
		aw:     opts.Awareness,
		status: StatusDisconnected,
		ctx:    ctx,
		cancel: cancel,
		done:   make(chan struct{}),
	}
	c.unsubs = []func(){
		c.doc.OnUpdate(c.handleDocUpdate),
		c.aw.OnUpdate(c.handleAwarenessUpdate),
		c.aw.OnChange(c.handleAwarenessChange),
	}
	return c, nil
}

// Status returns the current connection state.
func (c *Client) Status() Status {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.status
}

// Connect starts connecting (and reconnecting) in the background.
func (c *Client) Connect() {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.started || c.destroyed {
		return
	}
	c.started = true
	go c.run()
}

// Do runs fn while no remote update is being applied, so fn can read the
// document and then change it without a remote edit slipping in between.
func (c *Client) Do(fn func()) {
	c.applying.Lock()
	defer c.applying.Unlock()
	fn()
}

// Destroy announces departure, closes the connection and stops reconnecting.
func (c *Client) Destroy() {
	c.mu.Lock()
	if c.destroyed {
		c.mu.Unlock()
		return
	}
	c.mu.Unlock()

	// Queued on the current connection by handleAwarenessUpdate.
	c.aw.SetLocalState(nil)

	c.mu.Lock()
	c.destroyed = true
	out := c.out
	c.out = nil
	started := c.started
	c.mu.Unlock()

	for _, unsub := range c.unsubs {
		unsub()
	}
	if out != nil {
		out.drain()
	}
	c.cancel()
	if out != nil {
		_ = out.conn.Close()
	}
	if started {
		<-c.done
	}
	c.setStatus(StatusDisconnected)
}

func (c *Client) run() {
	defer close(c.done)
	attempts := 0
	for {
		c.setStatus(StatusConnecting)
		conn, err := c.opts.Dial(c.ctx, c.opts.URL, c.opts.Header)
		if err == nil {
			attempts = 0
			err = c.serve(conn)
		}
		if c.ctx.Err() != nil {
			return
		}
		c.dropRemoteAwareness()
		if closeCode(err) == RoomClosed {
			c.setStatus(StatusClosed)
			return
		}
		c.onError(err)
		c.setStatus(StatusDisconnected)

		backoff := min(c.opts.MaxBackoff, c.opts.MinBackoff<<min(attempts, 16))
		delay := time.Duration(float64(backoff) * (0.5 + rand.Float64()/2))
		attempts++
		select {
		case <-time.After(delay):
		case <-c.ctx.Done():
			return
		}
	}
}

// serve runs one connection until it fails.
func (c *Client) serve(conn Conn) error {
	out := newOutbox(conn)
	c.mu.Lock()
	if c.destroyed {
		c.mu.Unlock()
		_ = conn.Close()
		return c.ctx.Err()
	}
	c.out = out
	c.mu.Unlock()
	c.setStatus(StatusConnected)

	c.send(MessageSync, ysync.EncodeSyncStep1(c.doc))
	step2 := encoding.NewEncoder()
	step2.WriteVarUint(ysync.MsgSyncStep2)
	step2.WriteVarBytes(c.doc.EncodeStateAsUpdate())
	c.send(MessageSync, step2.Bytes())
	c.sendAwareness([]uint64{c.aw.ClientID()})

	for {
		data, err := conn.Read(c.ctx)
		if err != nil {
			c.mu.Lock()
			if c.out == out {
				c.out = nil
			}
			c.mu.Unlock()
			out.stop()
			_ = conn.Close()
			return err
		}
		if err := c.receive(data); err != nil {
			c.onError(err)
		}
	}
}

func (c *Client) receive(data []byte) error {
	plaintext, err := c.cipher.Decrypt(data)
	if err != nil {
		return err
	}
	t, payload, err := DecodeMessage(plaintext)
	if err != nil {
		return err
	}
	if t == MessageSync {
		c.applying.Lock()
		reply, err := ysync.ApplySyncMessage(c.doc, payload, c)
		c.applying.Unlock()
		if err != nil {
			return err
		}
		if len(reply) > 0 {
			c.send(MessageSync, reply)
		}
		return nil
	}
	update, err := encoding.NewDecoder(payload).ReadVarBytes()
	if err != nil {
		return err
	}
	return c.aw.ApplyUpdate(update, c)
}

func (c *Client) handleDocUpdate(update []byte, origin any) {
	if origin == c {
		return
	}
	c.send(MessageSync, ysync.EncodeUpdate(update))
}

func (c *Client) handleAwarenessUpdate(ev awareness.UpdateEvent) {
	if ev.Origin == c {
		return
	}
	ids := append(append(append([]uint64{}, ev.Added...), ev.Updated...), ev.Removed...)
	c.sendAwareness(ids)
}

// Someone new showed up: tell them about us right away instead of waiting for the
// periodic awareness renewal.
func (c *Client) handleAwarenessChange(ev awareness.ChangeEvent) {
	if ev.Origin == c && len(ev.Added) > 0 {
		c.sendAwareness([]uint64{c.aw.ClientID()})
	}
}

func (c *Client) dropRemoteAwareness() {
	// Every remote client counts as expired after a zero timeout; the local one never does.
	c.aw.RemoveExpired(0)
}

func (c *Client) sendAwareness(ids []uint64) {
	enc := encoding.NewEncoder()
	enc.WriteVarBytes(c.aw.EncodeUpdate(ids))
	c.send(MessageAwareness, enc.Bytes())
}

// send queues a frame on the current connection, or drops it while
// disconnected: the handshake on reconnect carries everything that was missed.
func (c *Client) send(t MessageType, payload []byte) {
	c.mu.Lock()
	out := c.out
	c.mu.Unlock()
	if out == nil {
		return
	}
	out.push(c.cipher.Encrypt(EncodeMessage(t, payload)))
}

func (c *Client) setStatus(s Status) {
	c.mu.Lock()
	if c.status == s {
		c.mu.Unlock()
		return
	}
	c.status = s
	c.mu.Unlock()
	if c.opts.OnStatus != nil {
		c.opts.OnStatus(s)
	}
}

func (c *Client) onError(err error) {
	if err != nil && c.opts.OnError != nil {
		c.opts.OnError(err)
	}
}

// outbox writes frames to a connection in order, without blocking the sender.
type outbox struct {
	conn   Conn
	mu     sync.Mutex
	queue  [][]byte
	closed bool
	wake   chan struct{}
	done   chan struct{}
}

func newOutbox(conn Conn) *outbox {
	o := &outbox{conn: conn, wake: make(chan struct{}, 1), done: make(chan struct{})}
	go o.loop()
	return o
}

func (o *outbox) push(frame []byte) {
	o.mu.Lock()
	if o.closed {
		o.mu.Unlock()
		return
	}
	o.queue = append(o.queue, frame)
	o.mu.Unlock()
	o.signal()
}

func (o *outbox) signal() {
	select {
	case o.wake <- struct{}{}:
	default:
	}
}

func (o *outbox) loop() {
	defer close(o.done)
	for {
		o.mu.Lock()
		queue := o.queue
		o.queue = nil
		closed := o.closed
		o.mu.Unlock()
		for _, frame := range queue {
			if err := o.conn.Write(context.Background(), frame); err != nil {
				o.mu.Lock()
				o.closed = true
				o.queue = nil
				o.mu.Unlock()
				return
			}
		}
		if closed && len(queue) == 0 {
			return
		}
		if len(queue) == 0 {
			<-o.wake
		}
	}
}

// drain stops accepting frames and waits (for a while) until the queued ones are written.
func (o *outbox) drain() {
	o.mu.Lock()
	o.closed = true
	o.mu.Unlock()
	o.signal()
	select {
	case <-o.done:
	case <-time.After(5 * time.Second):
	}
}

// stop discards queued frames without waiting.
func (o *outbox) stop() {
	o.mu.Lock()
	o.closed = true
	o.queue = nil
	o.mu.Unlock()
	o.signal()
}
